File: packages/agent-cache-py/betterdb_agent_cache/cluster.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any

from .errors import ValkeyCommandError


async def cluster_scan(
    client: Any,
    pattern: str,
    on_keys: Callable[[list[str], Any], Awaitable[None]],
    count: int = 100,
) -> None:
    """Scan matching keys across all master nodes (cluster) or the single node
    (standalone), calling on_keys(keys, client) with each non-empty batch.

    The same client is passed to on_keys in both modes. valkey-py's cluster
    client routes individual single-key commands automatically, so callers can
    pipeline DEL/GET per-key without hitting CROSSSLOT errors.
    """
    try:
        from valkey.asyncio.cluster import ValkeyCluster

        is_cluster = isinstance(client, ValkeyCluster)
    except ImportError:
        is_cluster = False

    if is_cluster:
        nodes = client.get_primaries()
        if not nodes:
            raise ValkeyCommandError(
                "SCAN", Exception("cluster has no master nodes visible")
            )

        for node in nodes:
            cursor = 0
            while True:
                try:
                    cursor, keys = await client.scan(
                        cursor, match=pattern, count=count, target_nodes=node
                    )
                except Exception as exc:
                    raise ValkeyCommandError("SCAN", exc) from exc

                decoded = [k.decode() if isinstance(k, bytes) else k for k in keys]
                if decoded:
                    await on_keys(decoded, client)
                if cursor == 0:
                    break
    else:
        cursor = 0
        while True:
            try:
                cursor, keys = await client.scan(cursor, match=pattern, count=count)
            except Exception as exc:
                raise ValkeyCommandError("SCAN", exc) from exc

            decoded = [k.decode() if isinstance(k, bytes) else k for k in keys]
            if decoded:
                await on_keys(decoded, client)
            if cursor == 0:
                break

File: packages/semantic-cache/src/cluster.ts
import type Valkey from 'iovalkey';
import { Cluster } from 'iovalkey';
import { ValkeyCommandError } from './errors';

/**
 * True when the client talks to a cluster, so callers can pick a command shape
 * the cluster accepts. Multi-key commands are the usual reason to ask: a
 * cluster rejects them with CROSSSLOT unless every key hashes to one slot.
 */
export function isClusterClient(client: Valkey): boolean {
  return client instanceof Cluster;
}

function getMasterNodes(client: Valkey): Valkey[] {
  if (!(client instanceof Cluster)) return [client];
  // Cast needed: TypeScript can't narrow Valkey & Cluster because both classes
  // share a private `reconnectTimeout` field, collapsing the intersection to never.
  return (client as unknown as Cluster).nodes('master') as Valkey[];
}

/**
 * Perform a SCAN across all master nodes if the client is a Cluster instance,
 * or on the single client if standalone. Calls `onKeys` with each batch of
 * matched keys. The caller handles what to do with them (GET, DEL, etc.).
 *
 * This is the single place in the codebase that handles the cluster vs
 * standalone SCAN divergence.
 */
export async function clusterScan(
  client: Valkey,
  pattern: string,
  onKeys: (keys: string[], nodeClient: Valkey) => Promise<void>,
  count: number = 100,
): Promise<void> {
  const nodes = getMasterNodes(client);

  if (nodes.length === 0) {
    throw new ValkeyCommandError('SCAN', new Error('cluster has no master nodes visible'));
  }

  for (const nodeClient of nodes) {
    let cursor = '0';
    do {
      let scanResult: [string, string[]];
      try {
        scanResult = await nodeClient.scan(cursor, 'MATCH', pattern, 'COUNT', count);
      } catch (err) {
        throw new ValkeyCommandError('SCAN', err);
      }
      cursor = scanResult[0];
      const keys = scanResult[1];

      if (keys.length > 0) {
        await onKeys(keys, nodeClient);
      }
    } while (cursor !== '0');
  }
}
